feat(rocprim): output benchmark iteration info

MyNameIsTrez opened the pull request:
## Motivation

This PR adds an optional benchmark flag, which specifies the JSON path to output individual benchmark iteration times to:

```json
{
    "{\"lvl\":\"device\",\"algo\":\"find_first_of\",\"keys_size\":1,\"first_occurrence\":\"0.100000\",\"value_type\":\"int8_t\",\"cfg\":\"default_config\"}/manual_time": [
        0.09152,
        0.09184,
        0.092001
    ],
    "{\"lvl\":\"device\",\"algo\":\"find_first_of\",\"keys_size\":1,\"first_occurrence\":\"0.500000\",\"value_type\":\"int8_t\",\"cfg\":\"default_config\"}/manual_time": [
        0.388802,
        0.389442,
        0.389442
    ]
}
```

## Technical Details

This JSON data has already been used by StreamHPC's internal `plot_iteration_times.py` script to generate graphs like these:
<img width="1252" height="693" alt="image" src="https://github.com/user-attachments/assets/0b8d775a-f23f-40f2-a691-aadd3e2a7065" />

Visualizing the benchmark iteration times has helped diagnose why some benchmarks are very noisy.

I also added the commit `refactor(rocprim): stop shadowing the method name`, which just fixes an oversight of mine when I updated this benchmark a few months ago.

The reason the `output_iteration_times()` and `update_iteration_times()` functions this pull request adds are so complex, is because I didn't want to introduce a new dependency on a JSON library.

In a future PR called `Trim outlier iteration times` that is currently under internal review, I noticed that I forgot to add a description above `output_iteration_times()`, so that PR adds this missing description:
```c
// This function overwrites any specialization's previously output JSON array.
// This allows gbench to call this fn several times per specialization.
void output_iteration_times() const
{
```

## Test Plan

This pull request has been reviewed and used internally at StreamHPC.

### Setup
1. Clone rocm-libraries.
2. Checkout this PR.
3. Run `gfx="$(rocminfo | grep -o 'gfx[0-9|a-f]\+' | head -n1)"`.
4. Run `mkdir build && cd build && CXX=hipcc cmake -GNinja -DBUILD_BENCHMARK=ON -DBENCHMARK_CONFIG_TUNING=OFF -DBENCHMARK_AUTOTUNED_TYPES_ONLY=ON -DGPU_TARGETS="$gfx" ..`

### Running benchmark_device_find_first_of
There are two options:
1. Generate `times.json` by running `ninja benchmark_device_find_first_of && time benchmark/benchmark_device_find_first_of --seed 42 --benchmark_out=foo.json --iteration_times_out times.json`.
2. Generate `iteration_times/benchmark_device_find_first_of_gfx90a.json` by running `python3 .gitlab/run_benchmarks.py --benchmark_dir build_benchmark/benchmark --benchmark_gpu_architecture "$gfx" --benchmark_output_dir benchmark_results --benchmark_filename_regex '^benchmark_device_find_first_of' --seed 42 --iteration_times_output_dir iteration_times`.

## Test Result

This JSON data has been successfully used to generate graphs.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

MyNameIsTrez commented:
@amd-hsong I've pushed four more commits. The result is that AMD SMI info is now *also* dumped alongside the iteration info.

I have verified that both `BENCHMARK_USE_AMDSMI=ON` and `BENCHMARK_USE_AMDSMI=OFF` let all benchmarks compile.

Here is what the new JSON format looks like, where I've omitted a lot of fields from the `metrics` objects:

<details>
<summary>JSON</summary>

```json
{
    "context": {
        "amdsmi": {
            "product_name": "Navi 48 [RX 9070/9070 XT]",
            "amdsmi_version": "25.3.0",
            "amdsmi_metrics_version": {
                "format_revision": 1,
                "content_revision": 3
            },
            "clocks": {
                "df": {
                    "min_clk": 313,
                    "max_clk": 313
                },
                "mem": {
                    "min_clk": 96,
                    "max_clk": 1258
                },
                "soc": {
                    "min_clk": 417,
                    "max_clk": 1476
                },
                "sys": {
                    "min_clk": 500,
                    "max_clk": 2400
                }
            },
            "energy_resolution": 15.300000,
            "stats": {
                "metrics": {
                    "temperature_edge": 46,
                    "current_socclk": 1371,
                    "voltage_gfx": 811
                },
                "clocks": {
                    "df": 4294967295,
                    "mem": 1124,
                    "soc": 1371,
                    "sys": 2400
                },
                "vram_used_bytes": 1112965120
            }
        }
    },
    "benchmarks": [
        {
            "name": "{\"lvl\":\"device\",\"algo\":\"find_first_of\",\"keys_size\":1,\"first_occurrence\":\"0.100000\",\"value_type\":\"int8_t\",\"cfg\":\"default_config\"}/manual_time",
            "batches": [
                {
                    "iterations_ms": [
                        0.075601,
                        0.0724
                    ],
                    "amdsmi_stats_after_iterations": {
                        "metrics": {
                            "temperature_edge": 46,
                            "current_socclk": 1371,
                            "voltage_gfx": 811
                        },
                        "clocks": {
                            "df": 4294967295,
                            "mem": 1124,
                            "soc": 1371,
                            "sys": 2379
                        },
                        "vram_used_bytes": 1548840960
                    }
                },
                {
                    "iterations_ms": [
                        0.07876,
                        0.074281
                    ],
                    "amdsmi_stats_after_iterations": {
                        "metrics": {
                            "temperature_edge": 46,
                            "average_socclk_frequency": 65535,
                            "voltage_gfx": 808
                        },
                        "clocks": {
                            "df": 4294967295,
                            "mem": 1124,
                            "soc": 1371,
                            "sys": 2374
                        },
                        "vram_used_bytes": 1548840960
                    }
                }
            ]
        }
    ]
}
```

</details>

amd-hsong reviewed (dismissed):
Looks good. @MyNameIsTrez There is a merge conflict to be resolved.

MyNameIsTrez commented:
@amd-hsong Thanks, I've rebased it onto develop and hopefully resolved the merge conflict. I also added a refactor commit.

It turns out that collecting the iteration times has negligible overhead, while collecting AMD SMI information makes the one benchmark I checked take 3x longer to run. So this is why I have disabled AMD SMI usage by default.
